### ISSUE TITLE
F# and Forth clash

### ISSUE THREAD

**Issue by ShalokShalom** [alecthomas/chroma#727]
**Describe the bug**

Chroma seems to handle both Forth and FSharp via .fs files. 
I get a clash with FSharp on both Codeberg and a private Gitea instance. 

Playground, by specifically selecting FSharp

![](https://user-images.githubusercontent.com/6344099/212427407-4a28a9cf-fd39-443b-8ec5-7b811bda429d.png)

Gitea

![](https://user-images.githubusercontent.com/6344099/212427523-3e3e035b-2b4b-4b2f-8417-45e1e2b95079.png)

It works with fsi and fsx files. (Other FSharp files)

**To Reproduce**

1. with .fs https://codeberg.org/ShalokShalom/FSKalc/src/branch/master/Program.fs
2. with .fsi https://try.gitea.io/jolheiser/FSKalc/src/branch/master/Program.fsi

Do *NOT* provide configuration for another tool (eg. Hugo). My time is limited and if you want me to fix your issue, help me help you.

**Comment by jolheiser** [comment]
Something to note, the `.fs` file looks correct in the `try.gitea.io` repo because of the `.gitattributes`, however the override also causes the repo language stats to change from `F#` to `FSharp`.

**Comment by alecthomas** [comment]
What's the proposed solution here?

**Comment by alecthomas** [comment]
Chroma inherited this from Pygments BTW, and it seems like the file extension is valid for both languages, so it's not clear what Chroma should do to correct this. My gut feeling is the user of Chroma should allow this to be configurable.

**Comment by ShalokShalom** [comment]
And a Gitea instance, that hosts bith Forth and F# files?

**Comment by alecthomas** [comment]
Well, like I said, what's the proposed solution? I understand what you're saying, but I do not see a solution.

**Comment by ShalokShalom** [comment]
Idk. Could Chroma be smart enough to read the file and figure out, if its one or the other? 

Like by comparing keywords. FSharp uses a very different syntax, at a first glance.

Could we do it this way?

**Comment by alecthomas** [comment]
Chroma has the concept of analysers, which are code fragments that use heuristics to detect a language. We could write an analyser for the two languages, but Gitea would have to be modified to use it (unless it already is).

**Comment by ShalokShalom** [comment]
I am fine with the Gitea part.

The analyzer part sounds like, its enough to write one, so if its not x, its y?

**Comment by alecthomas** [comment]
Yes, you can see how it's used [here](https://github.com/alecthomas/chroma#identifying-the-language). The lexer would need to implement the [Analyser](https://github.com/alecthomas/chroma/blob/master/lexer.go#L140) interface.

**Comment by alecthomas** [comment]
What we need is two representative source code samples that include some constructs that every example of that language would include, like differing comment types, etc.

**Comment by alecthomas** [comment]
[Here's](https://github.com/alecthomas/chroma/blob/master/lexers/dns.go#L11) an example of an Analyser.

**Comment by etareduction** [comment]
What's the workaround for now? What do i need to add in `.gitattributes` to fix highlighting in gitea?

**Comment by alecthomas** [comment]
We have support for `<analysis>` blocks in the XML now, so this should be solvable in the individual lexers by adding appropriate analyser expressions if someone wants to contribute some. Check the C lexer for an example.

**Comment by alecthomas** [comment]
I'm not sure if this is a problem still, but nobody has come forward to fix it, so I'm going to close this out.
